calico_containers/libnetwork_plugin/docker_plugin.py

```diff
@@ -109,96 +122,65 @@ def delete_network():
-        # We've failed to write the endpoint to the datastore.
-        # Back out the IP assignments and the veth creation.
-        app.logger.exception(e)
-        backout_ip_assignments(ep)
-        remove_veth(ep)
+    if next_hops.get(6):
+        ip6 = assign_ip("v6")
+        if ip6:
+            app.logger.info("Assigned IPv6 %s for ep %s" % (ip6, ep_id))
+        else:
+            app.logger.error("Failed to allocate IPv6 for endpoint %s", ep_id)
+    if not ip and not ip6:
+        app.logger.error("Failed to allocate and address for endpoint %s",
+                        ep_id)
         abort(500)
 
-    # Everything worked, create the JSON and return it to libnetwork.
-    assert len(ep.ipv4_nets) == 1
-    assert len(ep.ipv6_nets) <= 1
-    iface_json = {"ID": 0,
-                  "Address": str(list(ep.ipv4_nets)[0]),
-                  "MacAddress": ep.mac}
+    # Create the JSON to return to libnetwork
+    response = {"Interfaces":
+                    [{"ID": 0,
+                     "MacAddress": FIXED_MAC}]}
+    if ip:
+        response["Interfaces"][0]["Address"] = str(ip)
+    if ip6:
+        response["Interfaces"][0]["AddressIPv6"] = str(ip6)
 
-    if ep.ipv6_nets:
-        iface_json["AddressIPv6"] = str(list(ep.ipv6_nets)[0])
+    # Save this response along with the ep_id into the datastore.
+    client.write_cnm_endpoint(ep_id, response)
 
-    return jsonify({"Interfaces": [iface_json]})
+    return jsonify(response)
 
 
 @app.route('/NetworkDriver.DeleteEndpoint', methods=['POST'])
 def delete_endpoint():
     json_data = request.get_json(force=True)
+    app.logger.debug("DeleteEndpoint JSON=%s", json_data)
     ep_id = json_data["EndpointID"]
-    app.logger.info("Removing endpoint %s", ep_id)
-
-    # Remove the endpoint from the datastore, the IPs that were assigned to
-    # it and the veth. Even if one fails, try to do the others.
-    ep = None
-    try:
-        ep = client.get_endpoint(hostname=hostname,
-                                 orchestrator_id="docker",
-                                 workload_id=CONTAINER_NAME,
-                                 endpoint_id=ep_id)
-        backout_ip_assignments(ep)
-    except (KeyError, DataStoreError) as e:
-        app.logger.exception(e)
-        app.logger.warning("Failed to unassign IPs for endpoint %s", ep_id)
 
-    if ep:
-        try:
-            client.remove_endpoint(ep)
-        except DataStoreError as e:
-            app.logger.exception(e)
-            app.logger.warning("Failed to remove endpoint %s from datastore",
-                               ep_id)
+    # Backout IP assignment then remove CNM EP
+    cnm_ep = client.read_cnm_endpoint(ep_id)
 
-    # libnetwork expects us to delete the veth pair.  (Note that we only need
-    # to delete one end).
-    if ep:
-        remove_veth(ep)
+    if cnm_ep and client.delete_cnm_endpoint(ep_id):
+        app.logger.info("Removing endpoint %s", ep_id)
+        backout_ip_assignments(cnm_ep)
```

Comment:
Should we backout IP assignment first in case of failure mid way through?  If IP backout fails then there is no way to correlate endpoint with IPs again.